pkg/ipam/allocator.go

```diff
@@ -182,10 +184,36 @@ func (ipam *IPAM) AllocateNext(family, owner string) (ipv4Result, ipv6Result *Al
 	return
 }
 
-// ReleaseIP release a IP address.
-func (ipam *IPAM) ReleaseIP(ip net.IP) error {
-	ipam.allocatorMutex.Lock()
-	defer ipam.allocatorMutex.Unlock()
+// AllocateNextWithExpiration is identical to AllocateNext but registers an
+// expiration timer as well. This is identical to using AllocateNext() in
+// combination with StartExpirationTimer()
+func (ipam *IPAM) AllocateNextWithExpiration(family, owner string, timeout time.Duration) (ipv4Result, ipv6Result *AllocationResult, err error) {
+	ipv4Result, ipv6Result, err = ipam.AllocateNext(family, owner)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if timeout != time.Duration(0) {
+		for _, result := range []*AllocationResult{ipv4Result, ipv6Result} {
+			if result != nil {
+				defer func(result *AllocationResult) {
+					if err != nil {
+						ipam.ReleaseIP(result.IP)
+					}
+				}(result)
+
```

Comment:
This defer is confusing because it's not really needed since we are early returning on line 207.

Also, what happens if `ipam.StartExpirationTimer(result.IP, timeout)` fails for the `ipv4result`? When will `ipv6Result` be released?